### PR TITLE
support apex fused_rms_norm

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes : #896 

This PR adds a lookaside for `FusedRMSNormAffineMixedDtypesFunction.forward`  and corresponding symbol for the operation which is used by the lookaside. This  lookaside is enabled by default if the required import from `apex` (fused_layer_norm_cuda) is available. `apex_ex` (currently used for apex's cross-entropy) is re-used to register and implement this new symbol. So user has to also pass the `apex_ex` to the `executors` argument.

Also, we add compile option to disable `apex_ex` from claiming `cross_entropy` and newly added `FusedRMSNormAffineMixedDtypesFunction.forward`, so that using `apex_ex` doesn't make it mandatory to always use all the available apex operations.

NOTE - Lookaside is registered in `apex_fused_rms_norm.py` instead of usual `jit_ext.py` as this is operation may not be available always.

### CODE REVIEW

**Comment by crcrpar** [comment]
IMHO it'd be reasonable to rename & update https://github.com/Lightning-AI/lightning-thunder/blob/7433b861b2fe0747462ff35818180a1a57a6ad6f/thunder/executors/apex_entropyex.py for this, what's your take?

**Comment by kshitij12345** [comment]
> IMHO it'd be reasonable to rename & update https://github.com/Lightning-AI/lightning-thunder/blob/7433b861b2fe0747462ff35818180a1a57a6ad6f/thunder/executors/apex_entropyex.py for this, what's your take?

The reason I kept it seperate was that both these executors rely on different imports - entropyex relies on `xentropy_cuda` while rms_norm relies on `fused_layer_norm_cuda`. It is not necessary that user will always have both installed, so it would be odd (if we unify these under 1 executor) that this executor will only partially support certain operators depending on available packages. Because of this, I think we should keep them separate.

https://github.com/NVIDIA/apex/blob/59b80ee8df79cec125794949327f29913c328746/apex/normalization/fused_layer_norm.py#L69

**Comment by riccardofelluga** [comment]
I do support @crcrpar comment about updating the apex executor and keeping one single executor for apex. To check if an op is available or not we could use the checks like it's done in nvfuser executor. My main pain point on having two different imports for apex is that it can become pretty bad for practitioners to have to list all the different executors and give them as input to the jit call. We could also gate the ops behind compiler flag in the case the practitioners want to disable certain apex modules. For ref apex has table of different modules and i fear the day we have an executor for each row of this table 😱 https://github.com/NVIDIA/apex/tree/7b2e71b0d4013f8e2f9f1c8dd21980ff1d76f1b6?tab=readme-ov-file#custom-ccuda-extensions-and-install-options

**Review by crcrpar — COMMENTED** [review]
APEX has FusedRMSNorm `torch.autograd.Function`s and MixedFusedRMSNorm ones. IIRC/IIUC the latter ones were implemented for RMSNorm layers whose input and weight have different dtypes as in https://github.com/NVIDIA/apex/blob/59b80ee8df79cec125794949327f29913c328746/csrc/layer_norm_cuda.cpp#L358-L376.

Would you happen to know whether or not they are actually of different dtypes? If not, as mixed ones are used, we'd need to have some lookaside but it'd be possible to use non-mixed rms norm ones.

**Comment by kshitij12345** [comment]
> Would you happen to know whether or not they are actually of different dtypes? If not, as mixed ones are used, we'd need to have some lookaside but it'd be possible to use non-mixed rms norm ones.

Nemo uses the `MixedFusedRMSNorm` in `ParallelMLP` (Ref 1). I am not sure but depending on the configuration (with autocast, etc), it may happen that weights and inputs have different dtype.  So, I think we should just take care of MixedFusedRMSNorm for now. Wdyt?

NOTE - The weights are always initialized to `float` - Ref 2

Ref 1- https://github.com/tfogal/NeMo/blob/030360457ed1b51e6e5f15c2109c78cf5b76700d/nemo/collections/nlp/modules/common/megatron/mlp.py#L214-L218

Ref 2 - https://github.com/NVIDIA/apex/blob/59b80ee8df79cec125794949327f29913c328746/apex/normalization/fused_layer_norm.py#L401-L405

File: thunder/tests/test_apex_fused_norms.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+fused_layer_norm_cuda = pytest.importorskip("fused_layer_norm_cuda")
+from apex.normalization.fused_layer_norm import FusedRMSNormAffineMixedDtypesFunction
+from thunder.executors.apex_fused_rms_norm import register_apex_fused_rms_norm
+from thunder.executors.apex_entropyex import apex_ex
+import thunder
+
+
+@pytest.mark.parametrize("requires_grad", [True, False])
+@pytest.mark.parametrize("memory_efficient", [True, False])
+def test_apex_fused_norm(requires_grad, memory_efficient):
```
Comment (crcrpar) [inline thunder/tests/test_apex_fused_norms.py:14]: IMHO it'd be better if the name is e.g. `test_apex_fused_rms_norm` for clarity. what do you think?

Comment (kshitij12345) [reply thunder/tests/test_apex_fused_norms.py]: Agreed

File: docs/source/reference/executors/apex_entropyex.rst [hunk 2]
```diff
@@ -10,4 +10,4 @@ Executor of `NVIDIA/apex <https://github.com/nvidia/apex>`_.
 .. autosummary::
     :toctree: generated/
 
-    apex_available
```
Comment (kshitij12345) [inline docs/source/reference/executors/apex_entropyex.rst:13]: TODO - Update the name of this file.

**Review by crcrpar — APPROVED** [review]

File: thunder/executors/apex_fused_rms_norm_impl.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+)
+apex_ex.register_implementation(fused_rms_norm_backward, fused_rms_norm_backward)
+
+
+# Register the lookaside.
+def maybe_register_apex_fused_rms_norm_lookaside() -> None:
+    if APEX_FUSED_NORMS_AVAILABLE:
+
+        @thunder.core.jit_ext.register_general_jit_lookaside(FusedRMSNormAffineMixedDtypesFunction.forward)
+        @thunder.core.jit_ext.interpreter_needs_wrap
```
Comment (t-vi) [inline thunder/executors/apex_fused_rms_norm_impl.py:117]: Could this be handled by `replaces=FusedRMSNormAffineMixedDtypesFunction.forward` on the register_operator instead?

**Review by t-vi — COMMENTED** [review]
Looks great, thank you @kshitij12345 . I think the lookaside can be gotten easier (see comment).
I wonder whether we could abstract this to make the next thing easier (I imagine that the fused_rms_norm is quite typical of apex if not PyTorch extension operators in general). WDYT? (Though not necessarily in this PR.)  

Thank you @crcrpar for reviewing.

**Comment by kshitij12345** [comment]
> I wonder whether we could abstract this to make the next thing easier (I imagine that the fused_rms_norm is quite typical of apex if not PyTorch extension operators in general). WDYT? (Though not necessarily in this PR.)

I agree, making it easier would be good.

File: thunder/executors/apex_fused_rms_norm_impl.py (hunk 3 above)
Comment (t-vi) [reply thunder/executors/apex_fused_rms_norm_impl.py:117]: Seems good to me.

**Review by t-vi — APPROVED** [review]
lgtm
